exploration/function-composition-part-1.md

```diff
@@ -0,0 +1,824 @@
+at runtime:
+
+1. Resolved values can be bound to _variables_
+   (whose names arise from _declarations_ in the syntax).
+2. Resolved values can be passed to function implementations.
+3. Resolved values can be returned from function implementations.
+
+This implies:
+
+4. The resolved value returned from one function implementation
+can be passed to another function implementation,
+either one that implements the same function or a different function.
+
+which is the problem of composition.
+
+### Subsidiary problems
+
+  1. Define the runtime meaning of `.local` (and `.input`).
+
+> _In a declaration, the resolved value of the expression is bound to a variable, which is available for use by later expressions_
+(["Expression and Markup Resolution"](https://github.com/unicode-org/message-format-wg/blob/main/spec/formatting.md#expression-and-markup-resolution)).
+
+The term "resolved value" is left implementation-dependent, but its meaning affects the observable behavior of the formatter. (Note: Since `.input` can be treated as syntactic sugar for `.local`, we need only consider `.local` in examples.)
+
+  2. Define the types that custom functions manipulate.
+
+> _Call the function implementation with the following arguments... If the expression includes an operand, its resolved value.
+> ....If the call succeeds, resolve the value of the expression as the result of that function call._
+(["Function Resolution"](https://github.com/unicode-org/message-format-wg/blob/main/spec/formatting.md#function-resolution) )
+
+The same term, "resolved value", is used to describe the values
+that function implementations consume and produce.
+An implementation that supports custom functions needs to define concrete types
+(whose details depend on the underlying programming language of the implementation)
+that capture all the details of "resolved values".
+Implementors would benefit from guidance
+on how to map the MessageFormat concept of a "resolved value"
+onto a concrete type.
+
+  3. Define the semantics of composition
+
+A function can consume a value produced by another function,
+since the language provides `.local` declarations
+and any _variable_ can appear in an _expression_
+that has an _annotation_.
+(Functions implement _annotations_.)
+The meaning of this composition depends on the definition
+of "resolved value".
+Even with a precise definition of "resolved value",
+an additional question is raised of what the contract
+with custom functions needs to be
+in order to conform with the semantics of MessageFormat.
+Disambiguating the composability question
+affects observable behavior.
+
+  4. Provide useful names for different kinds of functions
+
+Currently, a given MessageFormat function can be thought of as a
+formatter, a selector, or both.
+Because the implementations for formatters and selectors
+naturally have different type signatures
+(a formatter consumes and produces a resolved value,
+while a selector produces a list of keys),
```

Comment:
I think this is a flaw in the current specification. I think it would be much more general for a selector to match keys, rather than produce them.

Reply:
As [currently specified](https://github.com/unicode-org/message-format-wg/blob/main/spec/formatting.md#resolve-preferences), keys are indeed "matched", but the result of that match is a preferential list of keys.

It is also possible to simplify this API in an implementation so that the resolved value of the selector is expected to have a `selectKey(keys)` method that returns a single key from its input list, rather than a preferential ordering, though the selection algorithm then gets a bit more complicated as it needs to track additional state (as done [here](https://github.com/messageformat/messageformat/blob/main/packages/mf2-messageformat/src/select-pattern.ts)). This works because the method can be called multiple times, each time leaving out the previously-best key from the input, to determine the full preference order.